rfcs/2021-02-bundle-scoping/proposal.md

```diff
@@ -0,0 +1,173 @@
+
+A scope would be a container for config variables only initially - other items such as containers and tasks would continue to exist globally. Config variables would only be able to be used in [expressions](https://batect.dev/docs/reference/config/expressions) in the same scope they are declared. For example, a config variable declared in the project's main configuration file could only be referenced in that file or in other files included in the project, but not from within any bundles added to the project.
+
+It would be valid to use the same name for config variables in different scopes. For example, `warnings_as_errors` could be defined in both the root project and a bundle, and these would be considered different variables, accessible from their respective scopes only.
+
+To better delineate between files that are included into the scope and bundles which have their own scope, the [`include`](https://batect.dev/docs/reference/config/includes#definition) property in configuration files would only be used for file includes, with a new `bundles` property added to support file-based or Git-based bundles. The existing behaviour for `include` (including support for Git includes and the use of global scope) would be preserved for a transition period, but any use of Git includes from `include` would generate a warning and eventually be removed.
+
+**Before:**
+
+```yaml
+include:
+  - tests.yml
+
+  - type: git
+    repo: https://github.com/batect/shellcheck-bundle.git
+    ref: 0.7.0
+
+  - type: git
+    repo: https://github.com/batect/hadolint-bundle.git
+    ref: 0.16.0
+
+  - bundles/my-cool-bundle/batect-bundle.yml
+```
+
+**After:**
+
+```yaml
+include:
+  - tests.yml
+
+bundles:
+  - type: git
+    repo: https://github.com/batect/shellcheck-bundle.git
+    ref: 0.7.0
+
+  - type: git
+    repo: https://github.com/batect/hadolint-bundle.git
+    ref: 0.16.0
+
+  - type: file
+    path: bundles/my-cool-bundle/batect-bundle.yml
+```
+
+
+
+To allow passing configuration to bundles, `bundles` would also allow setting values for bundles' config variables, for example:
+
+```yaml
+bundles:
+  - type: git
+    repo: https://github.com/batect/shellcheck-bundle.git
+    ref: 0.7.0
+    config:
+      warnings_as_errors: true
+      files: **/*.sh
+```
+
+These values could be literals or expressions. [Default values](https://batect.dev/docs/reference/config/config-variables#default) for config variables would also be extended to support expressions. Instantiating a bundle that has a config variable with no default without providing a value for that config variable would generate an error.
+
+
+
+## Stage 2: private containers
```

Comment:
Suggestion: I love the idea of private containers. Thoughts on having private tasks too?

Reply:
I considered it, but couldn't think of any concrete use cases for private tasks. Is there a use case you have in mind that would benefit from private tasks?

Reply:
The use case I was thinking of was where you might have prerequisites tasks for your own task. 

The exact use case is:

1. Publish DB Task
        - Start sql container + run command to deploy latest schema (binary) to the sql container
2. Run DB Task
        - PreReq = Public DB Task
        - Start sql container + run command "Press any key to continue command" 

In this example I don't really want to expose the first task. Since the functionality this bundle provides is to take a binary and run it on a DB container. 

(not sure if this can be done without multiple tasks)

Reply:
Right, makes sense. I think that would definitely fall into a "stage 3" in terms of implementation priority.

Reply:
Sounds good to me. 

To be completely honest it's not something that I "need". Just a nice to have so feel free to deprioritize it 🙂